### ISSUE TITLE
Server::nextConnectionAsync() blocks after the server has been shutdown

### ISSUE THREAD

**Issue by azjezz** [hhvm/hsl-experimental#131]
reproducing the issue : 
```hack
namespace Foo;

use namespace HH\Lib\{TCP, Network};

require_once __DIR__.'/vendor/autoload.hack';

<<__EntryPoint>>
async function main(): Awaitable<void> {
  \Facebook\AutoloadMap\initialize();

  $server = await TCP\Server::createAsync(Network\IPProtocolVersion::IPV4, '127.0.0.1', 1337);
  $i = $server->nextConnectionAsync();
  print 'stopping'."\n";

  $server->stopListening();
  print 'stopped, awaiting'."\n";
  await $i; // blocks forever
  print 'done'."\n";
}
```

**Comment by azjezz** [comment]
I can confirm that #125 doesn't solve this issue.

**Comment by azjezz** [comment]
for a workaround, I'm currently doing this (which doesn't exactly what I would expect ): 
![image](https://user-images.githubusercontent.com/29315886/82000579-b6676180-9650-11ea-8322-22b0a8962475.png)

**Comment by fredemmott** [comment]
Removing the abstractions:

```
use namespace HH\Lib\_Private\_OS;
<<__EntryPoint>>
async function main(): Awaitable<void> {
  $sin = new _OS\sockaddr_in(0, _OS\INADDR_LOOPBACK);
  $server = _OS\socket($sin->sa_family, _OS\SOCK_STREAM, 0);
  _OS\bind($server, $sin);
  _OS\listen($server, 128);

  _OS\fcntl($server, _OS\F_SETFL, (_OS\fcntl($server ,_OS\F_GETFL) as int )| _OS\O_NONBLOCK);
  try {
    _OS\accept($server);
  } catch (_OS\ErrnoException $e) {
    invariant($e->getCode() === _OS\EAGAIN, 'unexpected errno');
  }


  \var_dump('creating awaitable');
  $poll = _OS\poll_async($server, \STREAM_AWAIT_READ, 0);
  \var_dump('closing');
  _OS\close($server);
  \var_dump('awaiting awaitable');
  await $poll;
}
```

**Comment by fredemmott** [comment]
@jano : it looks like this is basically undefined behavior in the C++ side given the threading; I think that HHVM should track the `FileAwait`'s associated with each FD/resource, and either:
1. make the close call fail if any are pending, and provide a way to wait for them all
2. make the close() call cancel the fileawaits  - the stream_await()\poll_async() calls should fail with STREAM_AWAIT_CLOSED

IMO (2) is the correct solution - does it sound sane? I think this would be similar to the timeout cancellations - something like `runInEventThreadBaseAndWait(for(const auto& fa : m_fileawaits): fa.setFinished(FileAwait::CLOSED) }` ?

**Comment by azjezz** [comment]
I would go with 2
> make the close() call cancel the fileawaits - the stream_await()\poll_async() calls should fail with STREAM_AWAIT_CLOSED

I discovered this while playing around with [nuxed/http-server](https://github.com/nuxed/http-server).

If the server encounters an unrecoverable error while handling the connection, and the error occurred somewhere where the error handler cannot catch it, the server shuts down all sockets, in hope that the pending connections stop, either by canceling or throwing an exception ( which will be ignored ), as the upcoming connection would probably face the same issue.

**Comment by azjezz** [comment]
solution 1 means it won't be possible to stop the sockets until all upcoming connections have been "handled", and they will probably not be as it will result in the same unrecoverable error. so having N sockets open per example, means after the unrecoverable error occurred, I would need to "handle" N more connections.

**Comment by fredemmott** [comment]
This can be worked around with `\HH\Asio\cancel()` (which is almost always a bad idea, and supported on very few awaitables - this is literally the first good use case of it I've came across):

```
use namespace HH\Lib\_Private\_OS;
<<__EntryPoint>>
async function main(): Awaitable<void> {
  $sin = new _OS\sockaddr_in(0, _OS\INADDR_LOOPBACK);
  $server = _OS\socket($sin->sa_family, _OS\SOCK_STREAM, 0);
  _OS\bind($server, $sin);
  _OS\listen($server, 128);

  _OS\fcntl($server, _OS\F_SETFL, (_OS\fcntl($server ,_OS\F_GETFL) as int )| _OS\O_NONBLOCK);
  try {
    _OS\accept($server);
  } catch (_OS\ErrnoException $e) {
    invariant($e->getCode() === _OS\EAGAIN, 'unexpected errno');
  }


  \var_dump('creating awaitable');
  $poll = _OS\poll_async($server, 0x6, 0);
  _OS\close($server);
  \HH\Asio\cancel($poll, new \Exception('closed'));
  \var_dump('awaiting');
  await $poll;
}
```

Filed an HHVM issue for real fix, but should be able to get a workaround here.

**Comment by fredemmott** [comment]
With this:

```
diff --git a/src/network/_Private/CancelablePoller.php b/src/network/_Private/CancelablePoller.php
new file mode 100644
index 0000000..9cc13c8
--- /dev/null
+++ b/src/network/_Private/CancelablePoller.php
@@ -0,0 +1,45 @@
+<?hh
+/*
+ *  Copyright (c) 2004-present, Facebook, Inc.
+ *  All rights reserved.
+ *
+ *  This source code is licensed under the MIT license found in the
+ *  LICENSE file in the root directory of this source tree.
+ *
+ */
+
+namespace HH\Lib\_Private\_Network;
+
+use namespace HH\Lib\OS;
+use namespace HH\Lib\_Private\_OS;
+
+final class CancelablePoller {
+  private int $nextID = 0;
+  private dict<int, Awaitable<mixed>> $polls = dict[];
+
+  public async function pollAsync(
+    OS\FileDescriptor $fd,
+    int $events,
+    int $timeout_ns,
+  ): Awaitable<int> {
+    $id = $this->nextID;
+    $this->nextID++;
+
+    $awaitable = _OS\poll_async($fd, $events, $timeout_ns);
+    $this->polls[$id] = $awaitable;
+    try {
+      return await $awaitable;
+    } catch (PollCancelledException $e) {
+      return $e->getResult();
+    } finally {
+      unset($this->polls[$id]);
+    }
+  }
+
+  public function cancelAll(int $result): void {
+    $ex = new PollCancelledException($result);
+    foreach ($this->polls as $poll) {
+      \HH\Asio\cancel($poll, $ex);
+    }
+  }
+}
diff --git a/src/network/_Private/PollCancelledException.php b/src/network/_Private/PollCancelledException.php
new file mode 100644
index 0000000..e4a6683
--- /dev/null
+++ b/src/network/_Private/PollCancelledException.php
@@ -0,0 +1,21 @@
+<?hh
+/*
+ *  Copyright (c) 2004-present, Facebook, Inc.
+ *  All rights reserved.
+ *
+ *  This source code is licensed under the MIT license found in the
+ *  LICENSE file in the root directory of this source tree.
+ *
+ */
+
+namespace HH\Lib\_Private\_Network;
+
+final class PollCancelledException extends \Exception {
+  public function __construct(private int $result) {
+    parent::__construct();
+  }
+
+  public function getResult(): int {
+    return $this->result;
+  }
+}
diff --git a/src/network/_Private/socket_accept_async.php b/src/network/_Private/socket_accept_async.php
index 21550aa..0ab0981 100644
--- a/src/network/_Private/socket_accept_async.php
+++ b/src/network/_Private/socket_accept_async.php
@@ -16,13 +16,24 @@ use namespace HH\Lib\_Private\_OS;
 /** Accept a socket connection, waiting if necessary */
 async function socket_accept_async(
   OS\FileDescriptor $server,
+  CancelablePoller $poller,
 ): Awaitable<OS\FileDescriptor> {
   try {
     list($fd, $_addr) = OS\accept($server);
     return $fd;
   } catch (OS\BlockingIOException $_) {
     // accept (3P) defines select() as indicating the FD ready for read when there's a connection
-    await _OS\poll_async($server, \STREAM_AWAIT_READ, /* timeout = */ 0);
+    $result = await $poller->pollAsync(
+      $server,
+      \STREAM_AWAIT_READ, /* timeout = */
+      0,
+    );
+    if ($result == \STREAM_AWAIT_CLOSED) {
+      _OS\throw_errno(
+        OS\Errno::ECONNABORTED,
+        "Server socket closed while waiting for connection",
+      );
+    }
     list($fd, $_addr) = OS\accept($server);
     return $fd;
   }
diff --git a/src/tcp/Server.php b/src/tcp/Server.php
index 295e754..3dcbf4e 100644
--- a/src/tcp/Server.php
+++ b/src/tcp/Server.php
@@ -17,7 +17,10 @@ final class Server implements Network\Server<CloseableSocket> {
   /** Host and port */
   const type TAddress = (string, int);
 
+  private _Network\CancelablePoller $poller;
+
   private function __construct(private OS\FileDescriptor $impl) {
+    $this->poller = new _Network\CancelablePoller();
   }
 
   /** Create a bound and listening instance */
@@ -106,7 +109,7 @@ final class Server implements Network\Server<CloseableSocket> {
   }
 
   public async function nextConnectionAsync(): Awaitable<CloseableSocket> {
-    return await _Network\socket_accept_async($this->impl)
+    return await _Network\socket_accept_async($this->impl, $this->poller)
       |> new _TCP\CloseableTCPSocket($$);
   }
 
@@ -133,5 +136,6 @@ final class Server implements Network\Server<CloseableSocket> {
 
   public function stopListening(): void {
     OS\close($this->impl);
+    $this->poller->cancelAll(\STREAM_AWAIT_CLOSED);
   }
 }
diff --git a/src/unix/Server.php b/src/unix/Server.php
index cd6336f..554446c 100644
--- a/src/unix/Server.php
+++ b/src/unix/Server.php
@@ -17,7 +17,10 @@ final class Server implements Network\Server<CloseableSocket> {
   /** Path */
   const type TAddress = string;
 
+  private _Network\CancelablePoller $poller;
+
   private function __construct(private OS\FileDescriptor $impl) {
+    $this->poller = new _Network\CancelablePoller();
   }
 
   /** Create a bound and listening instance */
@@ -34,7 +37,7 @@ final class Server implements Network\Server<CloseableSocket> {
   }
 
   public async function nextConnectionAsync(): Awaitable<CloseableSocket> {
-    return await _Network\socket_accept_async($this->impl)
+    return await _Network\socket_accept_async($this->impl, $this->poller)
       |> new _Unix\CloseableSocket($$);
   }
 
@@ -45,5 +48,6 @@ final class Server implements Network\Server<CloseableSocket> {
 
   public function stopListening(): void {
     OS\close($this->impl);
+    $this->poller->cancelAll(\STREAM_AWAIT_CLOSED);
   }
 }
diff --git a/tests/tcp/HSLTCPTest.php b/tests/tcp/HSLTCPTest.php
index 03e8825..db71f5c 100644
--- a/tests/tcp/HSLTCPTest.php
+++ b/tests/tcp/HSLTCPTest.php
@@ -181,4 +181,19 @@ final class HSLTCPTest extends HackTest {
     expect($client_recv)->toEqual("foo bar\n");
     expect($server_recv)->toEqual("hello, world!\n");
   }
+
+  public async function testCloseDuringAsyncAccept(): Awaitable<void> {
+    $s = await TCP\Server::createAsync(
+      IPProtocolVersion::IPV4,
+      '127.0.0.1',
+      0,
+    );
+    // intentionally not awaiting
+    $accept_awaitable = $s->nextConnectionAsync();
+    $s->stopListening();
+    $ex = expect(async () ==> await $accept_awaitable)->toThrow(
+      OS\ErrnoException::class,
+    );
+    expect($ex->getErrno())->toEqual(OS\Errno::ECONNABORTED);
+  }
 }
```

... your original repro fails with:

```
Fatal error: Uncaught exception 'HH\Lib\OS\ConnectionAbortedException' with message 'ECONNABORTED(53): Server socket closed while waiting for connection' in /Users/fredemmott/code/hsl-experimental/src/os/_Private/Errno.php:48
```

Will put that up as a stacked diff on top of the native work. Not adding to PR given GitHub PRs don't fit a 'stacked diff' model well.

**Comment by fredemmott** [comment]
Fix is FB D21770499
